File: study/ts/script/scene/view/exploreMap/EventAnswerCell.ts
const { ccclass, property } = cc._decorator;

import { Lang } from '../../../lang/Lang';
import { assert } from '../../../utils/GlobleFunc';
import { TypeConvertHelper } from '../../../utils/TypeConvertHelper';
import CommonIconTemplateWithBg from '../../../ui/component/CommonIconTemplateWithBg';
import CommonButtonLevel1Highlight from '../../../ui/component/CommonButtonLevel1Highlight';

@ccclass
export default class EventAnswerCell extends cc.Component {

    static INDEX_CHAR_MAP = ["A", "B", "C", "D"];

    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageHighLight: cc.Sprite = null;

    @property({
        type: CommonIconTemplateWithBg,
        visible: true
    })
    _imageHero: CommonIconTemplateWithBg = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textAnswer: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageRight: cc.Sprite = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageWrong: cc.Sprite = null;

    @property({
        type: CommonButtonLevel1Highlight,
        visible: true
    })
    _btnAnswer: CommonButtonLevel1Highlight = null;

    private _index;
    private _answerId;
    private _answerPicId;
    private _description;
    private _callback;

    setUp(answerData, index, callback) {
        this._index = index;
        this._answerId = answerData['answer' + (index + '_id')];
        this._answerPicId = answerData['answer' + (index + '_picture')];
        this._description = answerData['answer' + (index + '_description')];
        this._callback = callback;
      //assert((this._callback, 'call back should not be nil!!');
        this.node.name = 'EventAnswerCell' + this._index;

        this._imageHighLight.node.active = false;
        this._imageRight.node.active = false;
        this._imageWrong.node.active = false;
        this._textAnswer.string = this._description;
        this._btnAnswer.setString(Lang.get('explore_answer_btn_name', { index: EventAnswerCell.INDEX_CHAR_MAP[this._index - 1] }));
        this._setAnswerImage();
    }
    _setAnswerImage() {
        this._imageHero.initUI(TypeConvertHelper.TYPE_HERO, this._answerPicId, null);
        this._imageHero.setImageTemplateVisible(true);
    }
    public onBtnAnswerClick() {
        console.log('onBtnAnswerClick ', this._index);
        this._callback(this._index);
    }
    //设置是否正确
    setRight(isRight) {
        this._imageRight.node.active = isRight;
        this._imageWrong.node.active = !isRight;
    }
    disableAnswer() {
        this._btnAnswer.setEnabled(false);
    }

}

File: study/lua/src/app/scene/view/exploreMap/EventAnswerCell.lua
local ViewBase = require("app.ui.ViewBase")
local EventAnswerCell = class("EventAnswerCell", ViewBase)
local Hero = require("app.config.hero")
local HeroRes = require("app.config.hero_res")

EventAnswerCell.INDEX_CHAR_MAP = {"A", "B", "C", "D"}

function EventAnswerCell:ctor(answerData, index, callback)
    --问题详细
    self._index = index
    self._answerId = answerData["answer"..index.."_id"]
    self._answerPicId = answerData["answer"..index.."_picture"]
    self._description = answerData["answer"..index.."_description"]
    self._callback = callback

    dump(answerData)
    assert(self._callback, "call back should not be nil!!")
    --ui
    self._btnAnswer = nil   --底板&触摸相应按钮
    self._imageHighLight = nil  --选中高亮
    self._imageHero = nil   --英雄肖像
    self._textAnswer = nil  --答案
    self._imageRight = nil  --对勾
    self._imageWrong = nil  --xx

    local resource = {
		file = Path.getCSB("EventAnswerCell", "exploreMap"),
        binding = {
            _btnAnswer = {
				events = {{event = "touch", method = "_onBtnAnswerClick"}}
			},
        }
	}
    self:setName("EventAnswerCell"..self._index)
	EventAnswerCell.super.ctor(self, resource)
end

function EventAnswerCell:onCreate()
    self._imageHighLight:setVisible(false)
    self._imageRight:setVisible(false)
    self._imageWrong:setVisible(false)
    self._textAnswer:setString(self._description)
    self._btnAnswer:setString(Lang.get("explore_answer_btn_name", {index = EventAnswerCell.INDEX_CHAR_MAP[self._index]}))
    -- self._imageHero:ignoreContentAdaptWithSize(true)
    -- self._imageHero:setScale(0.35)
    self:_setAnswerImage()
end

function EventAnswerCell:onEnter()
end

function EventAnswerCell:onExit()
end

function EventAnswerCell:_setAnswerImage()
    local TypeConvertHelper = require("app.utils.TypeConvertHelper")
    self._imageHero:initUI( TypeConvertHelper.TYPE_HERO ,self._answerPicId)
    self._imageHero:setImageTemplateVisible(true)
    -- local heroResId = Hero.get(self._answerPicId).res_id
    -- local heroRes = HeroRes.get(heroResId).story_res
    -- self._imageHero:loadTexture(Path.getChatRoleRes(heroRes))
end

function EventAnswerCell:_onBtnAnswerClick()
    self._callback(self._index)
end

--设置是否正确
function EventAnswerCell:setRight(isRight)
    -- self._imageHighLight:setVisible(true)
    self._imageRight:setVisible(isRight)
    self._imageWrong:setVisible(not isRight)
end

function EventAnswerCell:disableAnswer()
    -- self._imageHighLight:setVisible(true)
    self._btnAnswer:setEnabled(false)
end

return EventAnswerCell
